### PR TITLE
check if unsigned numbers underflow

### PR DESCRIPTION
**Pull request by Jobhdez**
@Izaakwltn  Hopefully this is not an issue that I made another pr. I had git issues with the other one.

This solves #921.

This is the interaction I just had in this new pr:

```
(coalton-toplevel 
		(declare add-8bit (U8 -> U8 -> U8))
		(define (add-8bit x y)
		  (+ x y))
		(declare sub-8bit (U8 -> U8 -> U8))
		(define (sub-8bit x y)
		  (- x y))
		(declare mul-8bit (U8 -> U8 -> U8))
		(define (mul-8bit x y)
		  (* x y))
		(declare fint (Integer -> U8))
		(define (fint x)
		  (fromInt x)))
; No value
COALTON-USER> (coalton (add-8bit 2 3))
5
COALTON-USER> (coalton (add-8bit 3 -4))
; Evaluation aborted on #<SIMPLE-ERROR "Unsigned value underflowed 8 bits." {1007939573}>.
COALTON-USER> (coalton (sub-8bit 3 2))
1
COALTON-USER> (coalton (sub-8bit 3 5))
; Evaluation aborted on #<SIMPLE-ERROR "Unsigned value underflowed 8 bits." {100835E453}>.
COALTON-USER> (coalton (mul-8bit 3 4))
12
COALTON-USER> (coalton (mul-8bit 3 -4))
; Evaluation aborted on #<SIMPLE-ERROR "Unsigned value underflowed 8 bits." {1008C8C2B3}>.
COALTON-USER> (coalton (fint 4))
4
COALTON-USER> (coalton (fint -2))
; Evaluation aborted on #<SIMPLE-ERROR "Unsigned value underflowed 8 bits." {10017D9693}>.
```

Thanks

### CODE REVIEW

**Comment by Izaakwltn** [comment]
I tested it and this provides correct underflow errors.

I think it would make sense to use `cl:error` instead of `cl:cerror`, and remove the `continue-format-string` since there is no wrapping operation defined. Currently, if you select ` 0: [CONTINUE] Continue, wrapping around.` in the debugger, you get another error because it's still expecting a number and isn't getting one:
```
The value
  COMMON-LISP:NIL
is not of type
  (UNSIGNED-BYTE 16)
   [Condition of type COMMON-LISP:TYPE-ERROR]
```

Otherwise, this seems solid.

**Comment by Izaakwltn** [comment]
Looks good to me! I have no gripes. 

I think there should be a future issue to have unsigned numbers error on overflow too, but that can be a separate thing.

I'm going to approve it but I'll wait to see if anyone else wants to review it before merging.

**Review by Izaakwltn — APPROVED** [review]

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:164]: Issues with tabs. Please convert to the appropriate number of spaces.

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:171]: I think there should be a CERROR allowing wrapping behavior.

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:170]: (< x 0) ==> (minusp x)

(- 0 x) ==> (- x)

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:176]: b is guaranteed >= 0 for unsigned numbers

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:170]: b can never be negative for unsigned numbers

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:182]: unsigned multiplication can't underflow

**Comment by stylewarning** [comment]
(Incidentally this means we have no tests that actually look for correct overflow / underflow behavior.)

**Comment by Jobhdez** [comment]
@stylewarning thank you for this.

I will try to address your review as soon as I can. Thanks. 

For the wrapper, I just have to make it so when the unsigned values underflow, the user can continue, if they want, with the highest number possible of the given unsigned type right?

so if two addition of U8 integers underflow, the wrapper needs to allow the user continue from the error with the wrapped number which in this case would be 255? correct? ust trying to make sure I am thinking the right thing. thanks

File: library/math/num.lisp
Comment (Jobhdez) [reply library/math/num.lisp:182]: what do you mean by this? are you saying the code that i wrote doesnt detect the underflow or that in general unsigned multiplication doesnt underflow?

wouldnt this be an underflow: `2 * -3`?

Comment (Izaakwltn) [reply library/math/num.lisp:182]: -3 isn't an unsigned number

File: library/math/num.lisp
Comment (Izaakwltn) [inline library/math/num.lisp:192]: what is the point of (cl:+ 0 b)? doesn't it just evaluate to b?

Comment (Jobhdez) [reply library/math/num.lisp:192]: I can simplify this. Thanks

File: library/math/num.lisp
Comment (Jobhdez) [reply library/math/num.lisp:203]: Will make this change as well; thanks

**Comment by Jobhdez** [comment]
@Izaakwltn addressed your review

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:164]: i think this can be deleted

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:174]: these can be deleted

File: library/math/num.lisp
Comment (stylewarning) [inline library/math/num.lisp:260]: these are re-defining `Num` for types that already have it. Above, `define-num-wrapping` is already defining `-` and the like for each of the listed types.

**Review by stylewarning — CHANGES_REQUESTED** [review]
Thanks for your patience in review.

This PR needs to modify `define-num-wrapping`'s definition of `-` and `fromInt` only. Adding another `define` macro is not needed (and is actually redundant), and the other "handler" functionality is not all that useful outside of the scope of this change, so it can be eliminated.

**Comment by Jobhdez** [comment]
> Thanks for your patience in review.
> 
> This PR needs to modify `define-num-wrapping`'s definition of `-` and `fromInt` only. Adding another `define` macro is not needed (and is actually redundant), and the other "handler" functionality is not all that useful outside of the scope of this change, so it can be eliminated.

mybad for the late reply. But your comment means that I only pretty much have take what I wrote in my function and move to the `define-num-wrapping` function? thanks

**Comment by stylewarning** [comment]
Yes, essentially.

**Comment by Jobhdez** [comment]
> Yes, essentially.

Ok I’ll get it done ASAP.

**Comment by Izaakwltn** [comment]
At least one test failing due to the `ecase`, specifically in test `iter-min-max`: 

```
While evaluating the form starting at line 3, column 0
  of #P"/__w/coalton/coalton/run-tests.lisp":
Unhandled SB-KERNEL:CASE-FAILURE in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                              {1001090073}>:
  62 fell through COMMON-LISP:ECASE expression. Wanted one of (8 16 32 64).
``` 

Your `ecase` can get matches for bits that aren't 8, 16, 32, or 64. 

I would recommend loading the coalton test suite and either running all tests locally or just running `iter-min-max` to debug.

**Comment by Izaakwltn** [comment]
I think `+unsigned-fixnum-bits+` is sbcl specific, so the tests are passing in SBCL but still not in the other implementations:

Allegro:
https://github.com/coalton-lang/coalton/actions/runs/8151586516/job/22279665406?pr=1025#step:5:507

CCL:
https://github.com/coalton-lang/coalton/actions/runs/8151586516/job/22279665807?pr=1025#step:6:434

You might be able to do an `#+sbcl` exception for that approach and find another way to handle the other implementations, but if you find a uniform solution for all three types that would be optimal.

**Comment by Jobhdez** [comment]
@Izaakwltn @stylewarning hey guys, my apologies for the delay on this issue. Im going to try izaak's suggestions today or tomorrow and figure out the cause. i do plan to finish this.

**Comment by stylewarning** [comment]
Due to the age of this PR, I'm going to be closing it. Please feel free to open another PR when you believe you've solved the issue. If the issue needs more clarification, please ask in the issue.

**Pull request closed without merging**
